crates/ark/src/lsp/diagnostics_syntactic.rs

```diff
@@ -0,0 +1,559 @@
+//
+// diagnostics_syntactic.rs
+//
+// Copyright (C) 2024 Posit Software, PBC. All rights reserved.
+//
+//
+
+use tower_lsp::lsp_types::Diagnostic;
+use tree_sitter::Node;
+use tree_sitter::Range;
+
+use crate::lsp::diagnostics::DiagnosticContext;
+use crate::lsp::encoding::convert_tree_sitter_range_to_lsp_range;
+use crate::lsp::traits::rope::RopeExt;
+use crate::treesitter::node_has_error;
+use crate::treesitter::NodeType;
+use crate::treesitter::NodeTypeExt;
+
+pub(crate) fn syntax_diagnostics(
+    root: Node,
+    context: &DiagnosticContext,
+) -> anyhow::Result<Vec<Diagnostic>> {
+    let mut diagnostics = Vec::new();
+
+    recurse(root, context, &mut diagnostics)?;
+
+    Ok(diagnostics)
+}
+
+fn recurse(
+    node: Node,
+    context: &DiagnosticContext,
+    diagnostics: &mut Vec<Diagnostic>,
+) -> anyhow::Result<()> {
+    if !node_has_error(&node) {
+        // No `ERROR` and no `MISSING`
+        return Ok(());
+    }
+
+    // `ERROR` handling stops recursion
+    if node.is_error() {
+        return diagnose_error(node, context, diagnostics);
+    }
+
+    // Look for contextual `MISSING` issues based on the node type
+    diagnose_missing(node, context, diagnostics)?;
+
+    recurse_children(node, context, diagnostics)
+}
+
+fn recurse_children(
+    node: Node,
+    context: &DiagnosticContext,
+    diagnostics: &mut Vec<Diagnostic>,
+) -> anyhow::Result<()> {
+    let mut cursor = node.walk();
+
+    for child in node.children(&mut cursor) {
+        recurse(child, context, diagnostics)?;
```

Comment:
It's not a full traversal! It's `O(1)`
https://github.com/tree-sitter/tree-sitter/blob/4f0d463d49a3f1f2a6eaa75ecc284a5719aad94e/lib/src/node.c#L511

It maintains the error cost as it builds the tree so this is super cheap

Reply:
Oh maybe we can do this

```rust
fn recurse(
    node: Node,
    context: &DiagnosticContext,
    diagnostics: &mut Vec<Diagnostic>,
) -> anyhow::Result<()> {
    let any_child_errors = recurse_children(node, context, diagnostics)?;

    if !any_child_errors && node.is_error() {
        // Report self as syntax error
        return diagnose_error(node, context, diagnostics);
    }

    // Look for contextual `MISSING` issues based on the node type
    diagnose_missing(node, context, diagnostics)
}
```

i.e. the _first_ thing that `recurse()` does it traverse the node's children. That way we can tell if any child of the current node has ever reported an error. If so, then we never report ourselves.

I think that would dramatically simplify `diagnose_error()` too. It would now unconditionally report the error, because we know we don't have any children that are also errors.

Reply:
yep that's the structure I was trying to explain

Reply:
https://github.com/posit-dev/ark/pull/523/commits/1006e79e40d987384876ea0af54cb2b8b5ba3401